css/ampshared.css

```diff
@@ -272,6 +272,14 @@ i-amphtml-sizer {
   display: none;
 }
 
+/** Hide all text node children of non-container elements. */
+.i-amphtml-notbuilt:not(.i-amphtml-layout-container) ,
+[layout]:not([layout=container]):not(.i-amphtml-element)
+{
+  color: transparent !important; 
+  line-height: 0 !important;
```

Comment:
Will this change the size of the unbuilt element? Eg, we want the layout box of an amp-img to occupy a 300x200 size even before it's built, so there's minimal page shift. This would only really be noticeable on cache pages with transformed output, but that's our most important case.

Reply:
This will change the size of an unbuilt element **if** the size relies on any text children to contribute to the dimensions. 

For example `amp-fit-text` with child `"hello world"` might have a 300x200 sizer, but its total size could be 300x218 if the `hello world` has a default `line-height` of 18px. This change _intends_ to change that initial size to be 300x200 instead of 300x218 because the post-layout resolved size will be 300x200.

I just took a look at the specific `amp-img` case. It looks like text nodes have an undesired affect on the resolved dimensions. i.e.

```html
<amp-img src="https://amp.dev/static/samples/img/amp.jpg" width="1080" height="610" layout="responsive" alt="AMP">
</amp-img>
```

In a 400 width viewport, the responsive size is 400x226.

```html
<amp-img src="https://amp.dev/static/samples/img/amp.jpg" width="1080" height="610" layout="responsive" alt="AMP">
  hello world
</amp-img>
```

In a 400 width viewport, the responsive size is 400x242. The difference being the 16px line height of the "hello world". 

These are the two visual representations of the above snippets in order with the sizer of the "hello world" sample highlighted:
![image](https://user-images.githubusercontent.com/10456171/82492981-28a6ce80-9ab5-11ea-8370-004cc0fac0d8.png)

In this case the unbuilt element of the second example will be changed from 400x242 to be 400x226, but after it is built it will be 400x242 again because the line-height will take up its default space instead of `0`. I'd expect this to cause layout shift. I think this issue wrt placeholder was described here as well: https://github.com/ampproject/amphtml/pull/27736

But I guess the question is how often people put text inside an amp-img? And the problem may not be that this change causes the text size to not be considered before built, but possibly instead that the text size is considered at all after being built.

This introduces another data point though that this change is pretty volatile and does not have a clearcut impact on all the components uniformly...

Reply:
@jridgewell To clarify, this "unbreaks" the layout of unbuilt elements that have text content. Essentially, this is the same logic applied to text content as we do with our `not-build > * {display: none}` a couple of lines above.

Reply:
This brings up: don't we have a test for this kind of thing? Can we add? We'd simply want to create an unupgraded `<amp-x class="i-amphtml-notbuilt">abc</amp-x>` element and check that it's `offsetHeight == 0`. If there's no better place for this kind of test now, we can just place it in `text-layout.js`.

Reply:
I recently added `test/integration/test-css.js` for something similar, but I think a visual diff test is a better fit. 

You can simulate non-upgraded elements by making the extension script URL wrong. That's what I do in the test fixture for `test-css.js`:
```html
<script async custom-element="amp-iframe" src="intentionally_wrong"></script>
```

Reply:
Good idea.

Reply:
Thanks all for these suggestions. Added a set of visual tests. Here is [the percy run representing the status prior to this change](https://percy.io/ampproject/caroqliu/builds/5848545), where the red section is the background color for the `<i-amphtml-sizer>` element and the grey section is the background color for the unbuilt component. 

[This is a sample percy run](https://percy.io/ampproject/caroqliu/builds/5848620) which accounts for the CSS changes in this PR.